### PR TITLE
Support @FormUrlEncoded

### PR DESCRIPTION
**Pull request by lwj1994**
fix https://github.com/lejard-h/chopper/issues/577

* Support @FormUrlEncoded，
* auto add content-type application/x-www-form-urlencoded 
* set ket.toString and value.toString() for @Body / @Field / @FieldMap data,  if data is `Map` and  `Map` is not `Map<String, String>`

### CODE REVIEW

File: chopper_generator/lib/src/generator.dart
Comment (Guldem) [inline chopper_generator/lib/src/generator.dart:503]: Not sure if this is the nicest way of checking the types. It's possible to check inner type as far I know. See [Response type checker](https://github.com/lejard-h/chopper/blob/2249102880d84f21854475bd641f4d5ce336f0c5/chopper_generator/lib/src/generator.dart#L602).

 But doesn't `mapType.endsWith('String>')` also mean its non-nullable. So the if statement only needs to check if its a `isStringValueType`.

File: chopper_generator/lib/src/generator.dart
Comment (Guldem) [inline chopper_generator/lib/src/generator.dart:508]: `forEach` is not always preferred. See [lint rule](https://dart.dev/tools/linter-rules/avoid_function_literals_in_foreach_calls).

**Review by Guldem — CHANGES_REQUESTED** [review]

File: chopper/lib/src/request.dart
Comment (lwj1994) [reply chopper/lib/src/request.dart]: set the cleanBody type to Map<String, String> for formurlencode

File: chopper_generator/lib/src/generator.dart
Comment (lwj1994) [reply chopper_generator/lib/src/generator.dart]: removed

File: chopper_generator/lib/src/generator.dart
Comment (Guldem) [inline chopper_generator/lib/src/generator.dart:513]: Not sure if nullable need to be checked. When looking at the "old" `FormUrlEncodedConverter` it doesn't care about null values.

**Review by Guldem — CHANGES_REQUESTED** [review]

File: chopper_generator/lib/src/generator.dart
Comment (lwj1994) [reply chopper_generator/lib/src/generator.dart:513]: > Not sure if nullable need to be checked. When looking at the "old" `FormUrlEncodedConverter` it doesn't care about null values.

null values are meaningless in most scenarios, maybe add an optional bool switch？

File: chopper_generator/lib/src/generator.dart
Comment (lwj1994) [reply chopper_generator/lib/src/generator.dart:307]: > I think this should be in the `else` part?
it  find the Class type of the @Body annotation.  After the test, the result was correct

File: chopper_generator/lib/src/generator.dart
Comment (lwj1994) [reply chopper_generator/lib/src/generator.dart:488]: rewrited

File: chopper_generator/lib/src/generator.dart
Comment (Guldem) [inline chopper_generator/lib/src/generator.dart:327]: Doesn't the `fieldMap` variable also exposed the `DartType`? Via `
fieldMap.values.first.typeValue;`?

File: chopper_generator/lib/src/generator.dart
Comment (Guldem) [inline chopper_generator/lib/src/generator.dart:330]: Question: What if you add multiple `@FieldMap()` annotation in one call? Is that something that's possible? Or should we throw a warning?

File: chopper_generator/lib/src/generator.dart
Comment (Guldem) [inline chopper_generator/lib/src/generator.dart:726]: Nit: Maybe change name to `enableToString` or something more descriptive?

File: chopper/lib/src/annotations.dart
Comment (Guldem) [inline chopper/lib/src/annotations.dart:436]: Was this even supported?

**Review by Guldem — CHANGES_REQUESTED** [review]
Nice additions! 👍  Looking good only got a few small questions

File: chopper_generator/lib/src/generator.dart
Comment (lwj1994) [reply chopper_generator/lib/src/generator.dart:327]: > Doesn't the `fieldMap` variable also exposed the `DartType`? Via ` fieldMap.values.first.typeValue;`?

i tried it but get error: `FormatException: Not an instance of Type.`

```dart
        try{
          _logger.info("fieldMapType = ${fieldMap.values.first.typeValue.toString()}");
        } catch (e) {
          _logger.info("error fieldMapType  \n${e}");
        }
```

run and output :
```dart
[INFO] Chopper Generator:error fieldMapType
FormatException: Not an instance of Type.
```

File: chopper/lib/src/annotations.dart
Comment (lwj1994) [reply chopper/lib/src/annotations.dart:436]: There is no support for `List<Map<String, dynamic>>` , I think it was a mistake.

File: chopper_generator/lib/src/generator.dart
Comment (lwj1994) [reply chopper_generator/lib/src/generator.dart:327]: in line 515. ConstantReader receive the  Annotation,  it is not a DartType，so it will crash when call `typeValue`. 
``` dart
    for (final ParameterElement p in method.parameters) {
      final DartObject? a = _typeChecker(type).firstAnnotationOf(p);
      if (annotation != null && a != null) {
        throw Exception(
          'Too many $type annotation for \'${method.displayName}\'',
        );
      } else if (annotation == null && a != null) {
        annotation = a;
        name = p.displayName;
      }
    }

    return annotation == null ? {} : {name: ConstantReader(annotation)};
  }
```

Comment (Guldem) [reply chopper_generator/lib/src/generator.dart:327]: Ok was just wondering. Thanks for looking into it!

**Review by Guldem — APPROVED** [review]
LGTM! Would be nice if we could also update docs. But that can also be a different MR.

**Comment by lwj1994** [comment]
> LGTM! Would be nice if we could also update docs. But that can also be a different MR.

👌🏻. Leave it to me

**Review by Guldem — APPROVED** [review]
Thanks for adding docs 👍

**Review by techouse — CHANGES_REQUESTED** [review]
Thanks for this. 😇

I added a few comments.

File: chopper_generator/test/test_service.dart
Comment (techouse) [reply chopper_generator/test/test_service.dart:2]: Nevermind. Wrong collection LOL

**Review by techouse — APPROVED** [review]
LGTM

**Comment by techouse** [comment]
Thank you for your contribution @lwj1994

**Pull request merged**
